### PR TITLE
Improve space and lookup-ref performance

### PR DESCRIPTION
**Pull request by ncalexan**
@rnewman -- couple of perf optimizations for you.  The first two set groundwork for further space optimizations; I see much better post-import DB size on disk, although I haven't quantified it today so I don't have numbers to copy-paste right now.

The lookup-ref perf is much improved, and the tests are finally in place.

### CODE REVIEW

**Comment by rnewman** [comment]
FYI this needs a rebase, @ncalexan.

File: src/datomish/sqlite.cljc [hunk 1]
```diff
@@ -81,7 +81,8 @@
 
 (defn all-rows
   [db [sql & bindings :as rest]]
-  (reduce-rows db rest [] conj))
+  (go-pair
+    (persistent! (<? (reduce-rows db rest (transient []) conj!)))))
```
Comment (rnewman) [inline src/datomish/sqlite.cljc]: Curiosity: does this help measurably with perf, particularly in JS?

I worry about the code size implications of lots of `go-pair`. Is it significant?

File: src/datomish/sqlite_schema.cljc [hunk 2]
```diff
@@ -115,6 +97,31 @@
...
+           index_avet0 TINYINT, index_vaet0 TINYINT,
+           index_fulltext0 TINYINT,
+           unique_value0 TINYINT,
+           sv BLOB,
+           svalue_type_tag SMALLINT,
+           rid INTEGER,
+           e INTEGER, a SMALLINT, v BLOB, tx INTEGER, value_type_tag SMALLINT)")])
+
+(defn create-temp-tx-lookup-eavt-statement [idx-name table-name]
+  ;; Note that `id_tx_lookup_added` is created and dropped
```
Comment (rnewman) [inline src/datomish/sqlite_schema.cljc]: Update comment.

File: src/datomish/sqlite_schema.cljc [hunk 3]
```diff
@@ -115,6 +97,31 @@
...
+  ;; after insertion, which makes insertion slightly faster.
+  ;; Prevent overlapping transactions.  TODO: drop added0?
+  [(str "CREATE UNIQUE INDEX IF NOT EXISTS "
+        idx-name
```
Comment (rnewman) [inline src/datomish/sqlite_schema.cljc]: You should probably document that these two functions don't escape these names.

**Comment by rnewman** [comment]
Commit message on part 2 has a typo.

File: src/datomish/db.cljc [hunk 4]
```diff
@@ -63,6 +63,23 @@
 (defn id-literal? [x]
   (instance? TempId x))
 
+(defrecord LookupRef [a v])
+
+(defn lookup-ref
+  [a v]
+  (if (and
+        (or (keyword? a)
```
Comment (rnewman) [inline src/datomish/db.cljc]: I think you should `datoms-attribute-transform` (that is, `entid`) in this case, throwing if it's not found. No point in continuing down the road with a keyword, only to eventually find that we don't know the attribute anyway.

File: src/datomish/db.cljc [hunk 5]
```diff
@@ -634,6 +658,61 @@
...
+            qs
+            (map
+              (fn [chunk]
+                (cons
+                  ;; Query string.
+                  (apply str "WITH t(searchid, a, v, value_type_tag) AS (VALUES "
+                         (apply str (repeater (count chunk))) ;; TODO: join?
+                         ") SELECT t.searchid, d.e
+                           FROM t, datoms AS d
+                           WHERE d.index_avet IS NOT 0 AND d.a = t.a AND d.value_type_tag = t.value_type_tag AND d.v = t.v")
```
Comment (rnewman) [inline src/datomish/db.cljc]: If `index_avet` can be `0` or `1`, saying `IS 1` here is probably more efficient.

File: src/datomish/db.cljc [hunk 6]
```diff
@@ -634,6 +658,61 @@
...
+
+                  ;; Bindings.
+                  (mapcat (fn [[[a v] searchid]]
+                            (let [a (entid db a)
```
Comment (rnewman) [inline src/datomish/db.cljc]: `entid` passes through keywords, which is probably not what we want.

File: src/datomish/db.cljc [hunk 7]
```diff
@@ -634,6 +658,61 @@
...
+              (partition-all (quot max-sql-vars 4) av->searchid))
+
+            ;; Map searchid -> e.  There's a generic reduce that takes [pair-chan] lurking in here.
+            searchid->e
+            (loop [coll {}
+                   qs qs]
+              (let [[q & qs] qs]
+                (if q
+                  (let [rs (<? (s/all-rows (:sqlite-connection db) q))
+                        coll* (into coll (map (juxt :searchid :e)) rs)]
```
Comment (rnewman) [inline src/datomish/db.cljc]: Bear in mind that `into` is just a transient `(reduce conj to from)` with a `persistent!` on the end. Here you're bouncing back through a persistent for each `q`.

You should inline that here:

```
(loop [coll (transient {})
      qs qs]
  (let [[q & qs] qs]
    (if q
      (let [rs (<? …)
          coll* (reduce conj! coll (map (juxt :searchid :e) rs))]
        (recur coll* qs))
      (persistent! coll))))
```

File: src/datomish/transact.cljc [hunk 8]
```diff
@@ -250,49 +250,40 @@
...
-                                              (or
-                                                ;; The lookup might fail! If so, throw.
-                                                (:e (<? (db/<av db a v)))
-                                                (raise "No entity found with attr " a " and val " v "."
-                                                       {:a a :v v}))
-                                              field))))))))))))))
+      (let [av->e    (<? (db/<avs db avs))
+            resolve1 (fn [field]
+                       (if-let [[a v] (->av (db/lookup-ref? field))]
+                         (if-not (unique-identity? (db/entid db a))
```
Comment (rnewman) [inline src/datomish/transact.cljc]: As mentioned, try to `entid` earlier.

File: src/datomish/transact.cljc [hunk 9]
```diff
@@ -250,49 +250,40 @@
...
 (defn <resolve-lookup-refs [db report]
   {:pre [(db/db? db) (report? report)]}
 
-  (let [entities (:entities report)]
-    ;; TODO: meta.
+  (let [unique-identity? (memoize (partial ds/unique-identity? (db/schema db)))
+        groups           (group-by (partial keep db/lookup-ref?) (:entities report))
+        entities         (get groups (lazy-seq)) ;; No lookup-refs? Pass through.
+        to-resolve       (dissoc groups (lazy-seq)) ;; The ones with lookup-refs.
+        avs              (set (map (juxt :a :v) (apply concat (keys to-resolve))))
```
Comment (rnewman) [inline src/datomish/transact.cljc]: Can you give some example ins and outs here? I think I know what's going on, but…

**Comment by rnewman** [comment]
I think a test that hits the splitting point, requiring multiple queries, would be valuable.

File: src/datomish/sqlite.cljc (hunk 1 above)
Comment (ncalexan) [inline src/datomish/sqlite.cljc]: My guess is not measurably.  `go-pair` is *definitely* hard on code size.  I'll drop this.

**Comment by ncalexan** [comment]
I've added a test that hits the splitting point.

**Comment by ncalexan** [comment]
Hmm I answered a few of your comments and nodded approval for others, but rebasing seems to have mostly lost them.  Hopefully they're in your email.  @rnewman, stamp?

**Comment by rnewman** [comment]
If it's faster, rebase on `origin/master` and bombs away!

**Comment by ncalexan** [comment]
> If it's faster, rebase on origin/master and bombs away!

Yeah -- 7 times faster for ~2000 lookup-refs.  See merge commit message.  Landed!

**Pull request closed without merging**